File: libs/scm/src/lib/core/services/dashboard.service.ts
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { BaseApiService } from './base-api.service';

export interface KpiCard {
  id: string;
  label: string;
  value: number;
  count?: number;
  total?: number;
  percentage?: number;
  trend: number;
  trendDirection: 'up' | 'down';
  breakdown: Record<string, number>;
}

export interface PipelineStage {
  stage: string;
  count: number;
  avgDays: number;
  targetDays: number;
  items: PipelineItem[];
}

export interface PipelineItem {
  id: string;
  ref: string;
  title: string;
  days: number;
  status: 'on_track' | 'near_target' | 'exceeding';
}

export interface TurnaroundMetric {
  label: string;
  actual: number;
  target: number;
  status: 'within_target' | 'near_target' | 'exceeding_target';
}

export interface ComplianceComponent {
  label: string;
  score: number;
  weight: number;
  detail: string;
}

export interface ComplianceScore {
  overall: number;
  trend: number;
  components: ComplianceComponent[];
}

export interface RecentTransaction {
  date: string;
  ref: string;
  description: string;
  type: string;
  typeClass: string;
  amount: string;
  status: string;
  statusClass: string;
  route: string;
  params: Record<string, string>;
}

export interface ExecutiveDashboard {
  kpiCards: KpiCard[];
  lifecyclePipeline: PipelineStage[];
  turnaroundTimes: TurnaroundMetric[];
  complianceScore: ComplianceScore;
  recentTransactions: RecentTransaction[];
}

export interface UifwCategory {
  amount: number;
  count: number;
  status: string;
  items?: Array<{
    ref: string;
    description: string;
    amount: number;
    status: string;
    reportedDate: string;
    responsible: string;
  }>;
}

export interface ComplianceGate {
  regulation: string;
  description: string;
  status: 'compliant' | 'partial' | 'non_compliant';
  detail: string;
  actionRequired?: number;
}

export interface SodRule {
  rule: string;
  violations: number;
  status: 'clean' | 'violation';
  detail?: string;
}

export interface FraudRiskIndicator {
  indicator: string;
  severity: 'low' | 'medium' | 'high' | 'info' | 'critical';
  detail: string;
  recommendation: string;
}

export interface ComplianceDashboard {
  uifw: {
    unauthorized: UifwCategory;
    irregular: UifwCategory;
    fruitless: UifwCategory;
    cumulative: { amount: number; priorYears: boolean };
  };
  complianceGates: ComplianceGate[];
  segregationOfDuties: SodRule[];
  fraudRiskIndicators: FraudRiskIndicator[];
}

export interface WorkloadItem {
  type: string;
  ref: string;
  title: string;
  days: number;
  priority: 'low' | 'medium' | 'high';
  route: string;
  params: Record<string, string>;
}

export interface Bottleneck {
  stage: string;
  itemCount: number;
  avgDays: number;
  targetDays: number;
  severity: 'on_track' | 'warning' | 'critical';
  route: string;
  params: Record<string, string>;
}

export interface SupplierScore {
  name: string;
  score: number;
  delivery: number;
  quality: number;
  invoiceAccuracy: number;
  bbbeeLevel: number;
}

export interface BudgetCommitment {
  vote: string;
  budget: number;
  committed: number;
  actual: number;
  available: number;
  utilisation: number;
}

export interface SpendCategory {
  category: string;
  amount: number;
  percentage: number;
}

export interface OperationalDashboard {
  myWorkload: {
    pendingApprovals: number;
    myRequisitions: number;
    grnToProcess: number;
    invoicesToMatch: number;
    contractsExpiring: number;
    overdueItems: number;
    items: WorkloadItem[];
  };
  bottlenecks: Bottleneck[];
  supplierScorecard: {
    top5: SupplierScore[];
    bottom5: SupplierScore[];
  };
  budgetCommitments: BudgetCommitment[];
  spendAnalytics: {
    byCategory: SpendCategory[];
    byBbbeeLevel: Array<{ level: string; amount: number; percentage: number }>;
    localVsNational: { local: number; national: number };
    monthlyTrend: Array<{ month: string; amount: number }>;
  };
}

export interface AiInsight {
  id: number;
  type: string;
  severity: 'info' | 'warning' | 'critical';
  icon: string;
  title: string;
  message: string;
  timestamp: string;
  category?: string;
  legislationRef?: string;
  legislationDetail?: string;
  actionRoute?: string;
  actionParams?: Record<string, string>;
  recommendation?: string;
  financialImpact?: string;
  impact?: string;
  confidence?: number;
  aiGenerated?: boolean;
}

export interface AttentionTile {
  stage: string;
  count: number;
  value: number;
  oldestAge: number;
  slaBreaches: number;
  trend: 'up' | 'down' | 'stable';
  route: string;
}

export interface AgingHeatmap {
  stages: string[];
  buckets: string[];
  data: number[][];
}

export interface BlockingEntry {
  user: string;
  role: string;
  stage: string;
  itemsPending: number;
  totalValue: number;
  avgAge: number;
  oldestAge: number;
  slaBreaches: number;
  escalated: boolean;
}

export interface FunnelStage {
  stage: string;
  count: number;
  value: number;
  avgDaysInStage: number;
  targetDays: number;
  overSla: number;
  valueStuck: number;
}

export interface Escalation {
  id: string;
  ref: string;
  description: string;
  stage: string;
  daysOverSla: number;
  value: number;
  escalatedTo: string;
  escalationLevel: 'user' | 'manager' | 'cfo';
  timestamp: string;
  status: 'active' | 'acknowledged' | 'resolved';
}

export interface ControlTowerDashboard {
  attentionTiles: AttentionTile[];
  agingHeatmap: AgingHeatmap;
  blockingLeaderboard: BlockingEntry[];
  bottleneckFunnel: FunnelStage[];
  escalationPanel: Escalation[];
}

@Injectable({ providedIn: 'root' })
export class DashboardService {
  private api = inject(BaseApiService);

  getExecutiveDashboard(): Observable<ExecutiveDashboard> {
    return this.api.apiGet<ExecutiveDashboard>('/dashboard/executive');
  }

  getComplianceDashboard(): Observable<ComplianceDashboard> {
    return this.api.apiGet<ComplianceDashboard>('/dashboard/compliance');
  }

  getOperationalDashboard(): Observable<OperationalDashboard> {
    return this.api.apiGet<OperationalDashboard>('/dashboard/operational');
  }

  getAiInsights(): Observable<{ insights: AiInsight[] }> {
    return this.api.apiGet<{ insights: AiInsight[] }>('/dashboard/ai-insights');
  }

  getComplianceAiInsights(): Observable<{ insights: AiInsight[] }> {
    return this.api.apiGet<{ insights: AiInsight[] }>('/ai/insights/compliance');
  }

  getOperationalAiInsights(): Observable<{ insights: AiInsight[] }> {
    return this.api.apiGet<{ insights: AiInsight[] }>('/ai/insights/operational');
  }

  getControlTowerAiInsights(): Observable<{ insights: AiInsight[] }> {
    return this.api.apiGet<{ insights: AiInsight[] }>('/ai/insights/control-tower');
  }

  getInvoicePaymentAiInsights(): Observable<{ insights: AiInsight[] }> {
    return this.api.apiGet<{ insights: AiInsight[] }>('/ai/insights/invoice-payment');
  }

  getControlTowerDashboard(): Observable<ControlTowerDashboard> {
    return this.api.apiGet<ControlTowerDashboard>('/dashboard/control-tower');
  }

  getRequisitionAiInsights(): Observable<{ insights: AiInsight[] }> {
    return this.api.apiGet<{ insights: AiInsight[] }>('/ai/insights/requisitions');
  }

  getQuotationAiInsights(): Observable<{ insights: AiInsight[] }> {
    return this.api.apiGet<{ insights: AiInsight[] }>('/ai/insights/quotations');
  }

  getOrderAiInsights(): Observable<{ insights: AiInsight[] }> {
    return this.api.apiGet<{ insights: AiInsight[] }>('/ai/insights/orders');
  }

  getInformalTenderAiInsights(): Observable<{ insights: AiInsight[] }> {
    return this.api.apiGet<{ insights: AiInsight[] }>('/ai/insights/informal-tenders');
  }

  getFormalTenderAiInsights(): Observable<{ insights: AiInsight[] }> {
    return this.api.apiGet<{ insights: AiInsight[] }>('/ai/insights/formal-tenders');
  }
}

File: SCM-API/Services/DashboardService.cs
using SCM_API.Helpers;
using SCM_API.Repositories.Interfaces;
using SCM_API.Services.Interfaces;

namespace SCM_API.Services;

public class DashboardService : IDashboardService
{
    private readonly IDashboardRepository _repository;
    private readonly ILogger<DashboardService> _logger;

    public DashboardService(IDashboardRepository repository, ILogger<DashboardService> logger)
    {
        _repository = repository;
        _logger = logger;
    }

    public async Task<object> GetDashboardStatsAsync(string? financialYear)
    {
        var fy = financialYear ?? FinancialYearHelper.GetCurrentFinancialYear();
        try { return await _repository.GetDashboardStatsAsync(fy); }
        catch (Exception ex) { _logger.LogWarning(ex, "DB unavailable for dashboard stats, returning fallback"); return GetFallbackDashboardStats(fy); }
    }

    public async Task<object> GetExecutiveDashboardAsync(string? financialYear)
    {
        var fy = financialYear ?? FinancialYearHelper.GetCurrentFinancialYear();
        try { return await _repository.GetDashboardStatsAsync(fy); }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "DB unavailable for executive dashboard, returning fallback");
            return new
            {
                kpiCards = new[]
                {
                    new { id = "requisitions", label = "Active Requisitions", value = 0, count = 0, total = 0, percentage = 0, trend = 0, trendDirection = "up", breakdown = new Dictionary<string, int>() },
                    new { id = "orders", label = "Purchase Orders", value = 0, count = 0, total = 0, percentage = 0, trend = 0, trendDirection = "up", breakdown = new Dictionary<string, int>() },
                    new { id = "payments", label = "Payments Processed", value = 0, count = 0, total = 0, percentage = 0, trend = 0, trendDirection = "up", breakdown = new Dictionary<string, int>() },
                    new { id = "invoices", label = "Invoices Pending", value = 0, count = 0, total = 0, percentage = 0, trend = 0, trendDirection = "up", breakdown = new Dictionary<string, int>() },
                    new { id = "budget", label = "Budget Utilization", value = 0, count = 0, total = 0, percentage = 0, trend = 0, trendDirection = "up", breakdown = new Dictionary<string, int>() },
                    new { id = "compliance", label = "Compliance Score", value = 0, count = 0, total = 0, percentage = 0, trend = 0, trendDirection = "up", breakdown = new Dictionary<string, int>() }
                },
                lifecyclePipeline = Array.Empty<object>(),
                turnaroundTimes = Array.Empty<object>(),
                complianceScore = new { overall = 0, trend = 0, components = Array.Empty<object>() },
                recentTransactions = Array.Empty<object>()
            };
        }
    }

    public Task<object> GetComplianceDashboardAsync(string? financialYear)
    {
        return Task.FromResult<object>(new
        {
            uifw = new
            {
                unauthorized = new { amount = 0, count = 0, status = "None reported" },
                irregular = new { amount = 0, count = 0, status = "None reported" },
                fruitless = new { amount = 0, count = 0, status = "None reported" },
                cumulative = new { amount = 0, priorYears = false }
            },
            complianceGates = Array.Empty<object>(),
            segregationOfDuties = Array.Empty<object>(),
            fraudRiskIndicators = Array.Empty<object>()
        });
    }

    public Task<object> GetOperationalDashboardAsync(string? financialYear)
    {
        return Task.FromResult<object>(new
        {
            myWorkload = new
            {
                pendingApprovals = 0,
                myRequisitions = 0,
                grnToProcess = 0,
                invoicesToMatch = 0,
                contractsExpiring = 0,
                overdueItems = 0,
                items = Array.Empty<object>()
            },
            bottlenecks = Array.Empty<object>(),
            supplierScorecard = new { top5 = Array.Empty<object>(), bottom5 = Array.Empty<object>() },
            budgetCommitments = Array.Empty<object>(),
            spendAnalytics = new
            {
                byCategory = Array.Empty<object>(),
                byBbbeeLevel = Array.Empty<object>(),
                localVsNational = new { local = 0, national = 0 },
                monthlyTrend = Array.Empty<object>()
            }
        });
    }

    public Task<object> GetControlTowerDashboardAsync(string? financialYear)
    {
        return Task.FromResult<object>(new
        {
            attentionTiles = Array.Empty<object>(),
            agingHeatmap = new { stages = Array.Empty<string>(), buckets = Array.Empty<string>(), data = Array.Empty<int[]>() },
            blockingLeaderboard = Array.Empty<object>(),
            bottleneckFunnel = Array.Empty<object>(),
            escalationPanel = Array.Empty<object>()
        });
    }

    public Task<object> GetAiInsightsAsync(string? financialYear)
    {
        return Task.FromResult<object>(new { insights = Array.Empty<object>() });
    }

    public async Task<object> GetRequisitionStatsAsync(string? financialYear)
    {
        var fy = financialYear ?? FinancialYearHelper.GetCurrentFinancialYear();
        try { return await _repository.GetRequisitionStatsAsync(fy); }
        catch (Exception ex) { _logger.LogWarning(ex, "DB unavailable for requisition stats"); return new { financialYear = fy, total = 0, pending = 0, approved = 0 }; }
    }

    public async Task<object> GetOrderStatsAsync(string? financialYear)
    {
        var fy = financialYear ?? FinancialYearHelper.GetCurrentFinancialYear();
        try { return await _repository.GetOrderStatsAsync(fy); }
        catch (Exception ex) { _logger.LogWarning(ex, "DB unavailable for order stats"); return new { financialYear = fy, total = 0, pending = 0, completed = 0 }; }
    }

    public async Task<object> GetInvoiceStatsAsync(string? financialYear)
    {
        var fy = financialYear ?? FinancialYearHelper.GetCurrentFinancialYear();
        try { return await _repository.GetInvoiceStatsAsync(fy); }
        catch (Exception ex) { _logger.LogWarning(ex, "DB unavailable for invoice stats"); return new { financialYear = fy, total = 0, pending = 0, paid = 0 }; }
    }

    public async Task<object> GetPaymentStatsAsync(string? financialYear)
    {
        var fy = financialYear ?? FinancialYearHelper.GetCurrentFinancialYear();
        try { return await _repository.GetPaymentStatsAsync(fy); }
        catch (Exception ex) { _logger.LogWarning(ex, "DB unavailable for payment stats"); return new { financialYear = fy, total = 0, processed = 0 }; }
    }

    public async Task<object> GetTenderStatsAsync(string? financialYear)
    {
        var fy = financialYear ?? FinancialYearHelper.GetCurrentFinancialYear();
        try { return await _repository.GetTenderStatsAsync(fy); }
        catch (Exception ex) { _logger.LogWarning(ex, "DB unavailable for tender stats"); return new { financialYear = fy, total = 0, active = 0 }; }
    }

    public async Task<object> GetContractStatsAsync(string? financialYear)
    {
        var fy = financialYear ?? FinancialYearHelper.GetCurrentFinancialYear();
        try { return await _repository.GetContractStatsAsync(fy); }
        catch (Exception ex) { _logger.LogWarning(ex, "DB unavailable for contract stats"); return new { financialYear = fy, total = 0, active = 0 }; }
    }

    public async Task<object> GetVendorStatsAsync(string? financialYear)
    {
        var fy = financialYear ?? FinancialYearHelper.GetCurrentFinancialYear();
        try { return await _repository.GetVendorStatsAsync(fy); }
        catch (Exception ex) { _logger.LogWarning(ex, "DB unavailable for vendor stats"); return new { financialYear = fy, total = 0, active = 0 }; }
    }

    public async Task<object> GetBudgetUtilizationAsync(string? financialYear)
    {
        var fy = financialYear ?? FinancialYearHelper.GetCurrentFinancialYear();
        return new { FinancialYear = fy, TotalBudget = 0m, Utilized = 0m, Available = 0m, UtilizationPercentage = 0m };
    }

    public async Task<object> GetProcurementSummaryAsync(string? financialYear)
    {
        var fy = financialYear ?? FinancialYearHelper.GetCurrentFinancialYear();
        return new { FinancialYear = fy, TotalProcurement = 0, TotalValue = 0m };
    }

    private object GetFallbackDashboardStats(string fy) => new { financialYear = fy, totalRequisitions = 0, totalOrders = 0, totalInvoices = 0, totalPayments = 0 };
}
